### ISSUE TITLE
visualistion of the task structure of plugins (auto generation)

### ISSUE THREAD

**Issue by Cudok** [BIM2SIM/bim2sim#641]
The tool generates automatically a the code of a structure diagram.
A simple structure of the tasks of one plugin and its dependencies (reads, touches) should be generated. 
 
The tool is written in python. The result is code of the structured diagram with the format of mermaid-js.
So it is possible to include the generated code of the structure into the documentation.

**Comment by Cudok** [comment]
First try: use python-mermaid:
- https://pypi.org/project/python-mermaid/

**Comment by Cudok** [comment]
## manual generated structure plot, Example teaser plugin
```mermaid
flowchart TB

subgraph "task LoadIFC"
  tli["common.LoadIFC"]
  subgraph reads & touches
    direction LR
    rli[/"None"/]
    toli[\"ifc_files"\]
  end
  extli("reads the IFC files of one or
         multiple domains inside bim2sim")
end

subgraph "task CheckIfc"
  tci["common.CheckIfc"]
  subgraph reads & touches
    direction LR
    rci[/"ifc_files"/]
    toci[\"None"\]
  end
  extci("Check an IFC file, for a number of conditions
         (missing information, incorrect information, etc)
         that could lead on future tasks to fatal errors.
        ")
end

subgraph "task CreateElements"
  tce["common.CreateElements"]
  subgraph reads & touches
    direction LR
    rce[/"ifc_files"/]
    toce[\"ifc_files, elements"\]
  end
  extce("Create bim2sim elements based
         on information in IFC.

         Here no values will be read
        ")
end

subgraph "task CreateSpaceBoundaries"
  tcs["bps.CreateSpaceBoundaries"]
  subgraph reads & touches
    direction LR
    rcs[/"ifc_files, elements"/]
    tocs[\"space_boundaries"\]
  end
  extcs("Create space boundary elements from ifc.
        ")
end

subgraph "task FilterTZ"
  tftz["bps.FilterTZ"]
  subgraph reads & touches
    direction LR
    rftz[/"elements"/]
    toftz[\"tz_elements"\]
  end
  extftz("Filters the thermal zones for later usage and stores them in state.
        ")
end

subgraph "task ProcessSlabsRoofs"
  tpsr["bps.ProcessSlabsRoofs"]
  subgraph reads & touches
    direction LR
    rpsr[/"elements"/]
    topsr[\"elements"\]
  end
  extpsr("Handles decomposed roofs and wrong slab elements.
        ")
end

subgraph "task BindStoreys"
  tbs["common.BindStoreys"]
  subgraph reads & touches
    direction LR
    rbs[/"elements"/]
    tobs[\"elements"\]
  end
  extbs("Bind thermal_zones and elements to
          each floor/storey and vice versa
        ")
end

subgraph "task EnrichUseConditions"
  teuc["bps.EnrichUseConditions"]
  subgraph reads & touches
    direction LR
    reuc[/"tz_elements"/]
    toeuc[\"enriched_tz"\]
  end
  exteuc("Enriches Use Conditions of thermal zones
          based on decisions and translation of zone names
        ")
end

subgraph "task VerifyLayersMaterials"
  tvlm["bps.VerifyLayersMaterials"]
  subgraph reads & touches
    direction LR
    rvlm[/"elements"/]
    tovlm[\"invalid"\]
  end
  extvlm("Verifies if layers and materials and their properties are meaningful.
        ")
end

subgraph "task EnrichMaterial"
  tem["bps.EnrichMaterial"]
  subgraph reads & touches
    direction LR
    rem[/"elements, invalid"/]
    toem[\"elements"\]
  end
  extem("Enriches material properties that were recognized as invalid
         LOD.layers = Medium & Full
        ")
end

subgraph "task DisaggregationCreation"
  tdc["bps.DisaggregationCreation"]
  subgraph reads & touches
    direction LR
    rdc[/"elements"/]
    todc[\"disaggregations"\]
  end
  extdc("Disaggregates building elements based on their space boundaries.

         This task is needed to allow the later combination for thermal zones. If two
         thermal zones are combined to one, we might need to cut/disaggregate
         elements like walls into pieces that belong to the different zones.
        ")
end

subgraph "task CombineThermalZones"
  tctz[" bps.CombineThermalZones"]
  subgraph reads & touches
    direction LR
    rctz[/"tz_elements, elements"/]
    toctz[\"bounded_tz"\]
  end
  extctz("Combine thermal zones to reduce the amount of thermal zones.

          As the zoning of simulation models is a time-consuming task we decided to
          automate it with the tasks.
          This task will combine multiple thermal zones into one zone based on the
          criteria selected in the simulation type settings and the decisions made.
          We do this by giving the user multiple criteria to select from:
           * External/Internal
           * Orientation
           * Usage
           * Window to wall ratio
       ")
end

subgraph "task Weather"
  tw["common.Weather"]
  subgraph reads & touches
    direction LR
    rw[/"elements"/]
    tow[\"weather_file"\]
  end
  extw("Task to get the weather file for later simulation.
       ")
end

subgraph "task LoadLibrariesTEASER"
  tllt["LoadLibrariesTEASER"]
  subgraph reads & touches
    direction LR
    rllt[/"None"/]
    tollt[\"libraries"\]
  end
  extllt("Load AixLib library for export.
       ")
end

subgraph "task CreateTEASER"
  tct["teaser_task.CreateTEASER"]
  subgraph reads & touches
    direction LR
    rct[/"libraries, elements, weather_file"/]
    toct[\"teaser_prj, bldg_names, orig_heat_loads, orig_cool_loads, tz_mapping"\]
  end
  extct("Creates a TEASER project by using the found information from IFC.
       ")
end

subgraph "task ExportTEASER"
  tet["teaser_task.ExportTEASER"]
  subgraph reads & touches
    direction LR
    ret[/"teaser_prj"/]
    toet[\"teaser_prj"\]
  end
  extet("Exports the previous created TEASER project to Modelica.
       ")
end

subgraph "task SimulateModelEBCPy"
  tsme["teaser_task.SimulateModelEBCPy"]
  subgraph reads & touches
    direction LR
    rsme[/"bldg_names"/]
    tosme[\"teaser_mat_result_paths, sim_results_path"\]
  end
  extsme("None
       ")
end

subgraph "task CreateResultDF"
  tcrdf["teaser_task.CreateResultDF"]
  subgraph reads & touches
    direction LR
    rcrdf[/"teaser_mat_result_paths, sim_results_path, tz_mapping"/]
    tocrdf[\"df_finals"\]
  end
  extcrdf("This ITask creates a result dataframe for TEASER BEPS simulations.

    Args:
        teaser_mat_result_paths: path to simulation result file
        tz_mapping: dict with mapping between IFC space GUIDs and rooms/zones
    Returns:
        df_final: final dataframe that holds only relevant data, with generic
        `bim2sim` names and index in form of MM/DD-hh:mm:ss
       ")
end

subgraph "task PlotBEPSResults"
  tpbepsr["bps.PlotBEPSResults"]
  subgraph reads & touches
    direction LR
    rpbepsr[/"df_finals, sim_results_path, ifc_files"/]
    topbepsr[\"None"\]
  end
  extpbepsr("Plots the results for BEPS simulations.

     This holds pre configured functions to plot the results of the BEPS
     simulations with EnergyPlus or TEASER.

     Args:
         df_finals: dict of final results where key is the building name and
          value is the dataframe holding the results for this building
         sim_results_path: path where to store the plots (currently with
          simulation results, maybe change this? #TODO
         ifc_files: bim2sim IfcFileClass holding the ifcopenshell ifc instance
       ")
end

tli --> tci --> tce --> tcs --> tftz --> tpsr --> tbs --> teuc --> tvlm --> tem --> tdc --> tctz --> tw --> tllt --> tct --> tet --> tsme --> tcrdf --> tpbepsr

**Comment by Cudok** [comment]
## open questions

1. Should the tool work on code basis or should it be part of the object structure.
  - Object structure:
    - Pro:
      - more integrated into the structure of bim2sim in general
    - Con:
      - more effort to run the tool, because it is needed to setup a bim2sim project

**Comment by Cudok** [comment]
## Results for: Teaser, central state visualization
```mermaid

**Comment by Cudok** [comment]
## Results for: EnergyPlus, decentral state visualization
```mermaid

**Comment by Cudok** [comment]
## Questions
1. [ ] What kind of visualization you like more: central state or decentral state?
2. [ ] Do you have improvement proposals?

Pls add your answers.

**Comment by ShujiaoHee** [comment]
> ## Questions
> 1. [ ]  What kind of visualization you like more: central state or decentral state?
> 2. [ ]  Do you have improvement proposals?
> 
> @DaJansenGit , @veronikarichter , @ShujiaoHee Pls add your answers.

Hi, from my point of view, I think the decentral state is better, more readable when scrolling through the page.

**Comment by veronikarichter** [comment]
> > ## Questions
> > 
> > 1. [ ]   What kind of visualization you like more: central state or decentral state?
> > 2. [ ]   Do you have improvement proposals?
> > 
> > @DaJansenGit , @veronikarichter , @ShujiaoHee Pls add your answers.
> 
> Hi, from my point of view, I think the decentral state is better, more readable when scrolling through the page.

Yes, same for me. Decentral visualization is easier to read/follow

**Comment by Cudok** [comment]
# links to the function generation the mermaid code are missing in the documentation
- will be solved in the issue #535
